pkg/cmd/util/clientcmd/factory.go

```diff
@@ -774,3 +776,43 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 	c.clients[config.GroupVersion.String()] = client
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(d discovery.DiscoveryInterface, m meta.RESTMapper) ([]unversioned.GroupResource, error) {
+	set := sets.NewString()
+	all, err := d.ServerResources()
+	if err != nil {
+		return nil, err
+	}
+	for apiVersion, v := range all {
+		gv, err := unversioned.ParseGroupVersion(apiVersion)
+		if err != nil {
+			continue
+		}
+		for _, r := range v.APIResources {
+			// ignore subresources
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			// because discovery info doesn't tell us whether the object is virtual or not, perform a lookup
+			// by the kind for resource (which should be the canonical resource) and then verify that the reverse
+			// lookup (KindsFor) does not error.
+			if mapping, err := m.RESTMapping(unversioned.GroupKind{Group: gv.Group, Kind: r.Kind}, gv.Version); err == nil {
+				if _, err := m.KindsFor(mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource)); err == nil {
+					gr := unversioned.GroupResource{Group: mapping.GroupVersionKind.Group, Resource: mapping.Resource}
+					set.Insert(gr.String())
```

Comment:
I'm still not seeing why you don't just have `map[unversioned.GroupResource]bool` to keep the types correct.  I don't see a reason to convert through a string.